### ISSUE TITLE
Inferring return type(`List<Object>`) for a function that actually evaluates to `List<Enum>`.

### ISSUE THREAD

**Issue by hamsbrar** [dart-lang/language#3290]
[Edit by eernstg: This issue revealed that the analyzer and the CFE treat the superinterface graph of an enum declaration differently. It should be decided which treatment is correct, if any, and implementations should then use that approach.]

```dart
enum E { a, b }

mixin M on Enum {}

abstract interface class I {}

enum E1 with M implements I { a1 }

enum E2 with M implements I { a2 }

List<Enum> f(E e) => switch (e) {
      E.a => E1.values,
      E.b => E2.values,
    };

void main() {
  f(E.a).forEach(print);
}
```

1. There are no warnings from the analyzer.

2. Compiler throws the following error:
    ```sh
    Error: A value of type 'List<Object>' can't be returned from a function with return type 'List<Enum>'.
    - 'List' is from 'dart:core'.
    - 'Object' is from 'dart:core'.
    - 'Enum' is from 'dart:core'.
    List<Enum> f(E e) => switch (e) {
                        ^
    ```

3. Analyzer report an issue if I remove the type constraint from the mixin `M`. i.e if I use `mixim M {}` not `mixin M on Enum {}`.
    ```sh
    Analyzing t1.dart...                   0.2s

      error • t1.dart:11:22 • A value of type 'List<Object>' can't be returned from the function 'f' because it has a return type of 'List<Enum>'. •
              return_of_invalid_type

    1 issue found.
    ```

4. Analyzer report the same issue if either `E1` or `E2` stop mixing with `M`.

5. Example work as expected if `f` is inlined.

6. Example work as expected if either `E1` or `E2` stop implementing `I`.

7. Example work as expected if there's only one match in switch. i.e it works if I use `switch (e) { _ => E1.values }`.

8. Example work as expected if I use switch statement. i.e it works if I use `switch{ case E.a: return ...`.

```sh
> dart --version
Dart SDK version: 3.2.0-74.0.dev (dev) (Tue Aug 15 13:04:34 2023 -0700) on "linux_x64"
```
Actual example(where I'm having this issue) is bit large but do let me know if it's required. Also it could be true that I'm not using some of the language features as they are *intended*, and in that case it'd great to see some relevant warnings.

**Comment by lrhn** [comment]
This seems to be a repeat of https://github.com/dart-lang/sdk/issues/51730

There is no bug, every implementation behaves according to specification. They just don't agree on what the class hierarchy is for `enum` types, because the analyzer uses only information from the specification, and the front-end also knows about internal, private helper classes used by the backends.

And since the upper-bound operation depends on the *depth* of a type in the class hierarchy, that means they produce different results.

I'm moving this issue to the language repository, because we might have to specify which behavior is actually correct, or prioritize our work on a better upper-bound algorithm. The introduction of switch expressions have increased the pressure on the upper bound computations, so the existing flaws show up more often. (Vastly more often, based on issued filed.)
And the difference between Analyzer and CFE on **Up** of enum types also shows up more often, so it's gone from an annoyance to a liability.

(One possible small step: Remove inaccessible types from the sets before doing LUB on them. That would mean that the **UP** function behaves differently depending on which library it's used from, but that may also mean it can more easily be computed modularly, using only public summarizes of other modules. But no code outside of `dart:core` would see the special `_Enum` class in their upper-bound computations. That's definitely breaking, because it will change the **Up** of existing types.)

**Comment by eernstg** [comment]
This still sounds like the CFE should remove `_Enum` from the sets of superinterfaces that are used to find the least upper bound in that Dart 1 part of `UP`, and skip `_Enum` when path lengths are computed. I don't see why it wouldn't be a goal (a soft one, best effort) to ensure that it is impossible to detect that `_Enum` exists.

**Comment by lrhn** [comment]
We can do something special-cased for the `_Enum` type, but it's not going to be the only problem, just the currently most visible one.
We could also just make the analyzer know about `_Enum`, after all, it's job is to predict runtime behavior, so it should match the actual runtime code as well as possible.

We could probably also add some unnecessary superclasses to `_Enum`, to give it greater depth. Then it would *win* in every such comparison, and the **Up** here would be `_Enum`.  Every time.

Like:
```dart
import "dart:convert";
void main() {
  var r = DateTime.now().millisecondsSinceEpoch < 0;
  var l = r ? AsciiEncoder() : Latin1Encoder();
  print([l].runtimeType); // JSArray<_UnicodeSubsetEncoder>
}
```
Here we leak that two unrelated encoders have a shared superclass in their implementation. 
In a perfect world, nobody would ever know that.

**Comment by eernstg** [comment]
> Here we leak that two unrelated encoders have a shared superclass in their implementation.
In a perfect world, nobody would ever know that.

That's a different topic because that's regular Dart code. We might compute least upper bounds in such a way that private declarations in different libraries are always removed from the set of potential results. However, I doubt that we would actually do that. At least, we'd need to double check that we don't get results which are too weird if `UP(A, B)` can yield different results in different libraries, even though `A` resolves to the same type in all of them, and similarly for `B`. Alternatively, we need to consider the consequences if private declarations are always ignored by `UP` and others, even in the library where those private types are declared.

Anyway, that's one discussion. The discussion about `_Enum` is different, because `_Enum` and `Enum` have non-standard semantics, and `_Enum` is basically specified to not exist. That's the reason why I think implementations should strive to make `_Enum` unobservable.

**Comment by lrhn** [comment]
The `_Enum` class actually _is_ specified:

> **Superclass**: The superclass of `C` is an implementation-specific built-in class _EnumImpl_, with the mixins declared by `E` applied. (The _EnumImpl_ class may be the `Enum` class itself or it may be another class which extends or implements `Enum`, but as seen from a non-platform library the interface of _EnumImpl_ is the same as that of `Enum`, and its methods work as specified for `Enum` )

It's not specified which class the _EnumImpl_ is, or what its depth.
That is, it's not specified that it _isn't_ there, but only that it might or might not be. It's open to being implementation dependent.

That's just highly annoying when **Up** behaves differently depending on that implementation choice.

You might be right that the best solution to this problem in isolation is to _formally specify_ that the _EnumImpl_ class should be ignored completely for **Up** computation: not included in the superinterfaces and doesn't count in the depth computation, for depth, the implicit declared superclass of an `enum` decoration counts as being `Enum`.

But there is still the general problem that the behavior of the classes that a library exports, wrt. **Up**, depends on internal implementation details, which may make it a breaking change to add or remove a private superclass. Something which should be completely invisible, to anyone who cannot name the supertype.

Or, short: _depth_ is part of your public API, and it shouldn't be. The set of public supertypes it's part of your API, private ones shouldn't be, they should be implementation details.

**Comment by eernstg** [comment]
> It's open to being implementation dependent [whether or not `Enum` and `_Enum`/`EnumImpl` is the same class]

Right (and I should have looked it up ;-), but I think it would be useful to avoid that the type of an expression is implementation dependent, and we seem to agree on that:

> That's just highly annoying when **Up** behaves differently depending on that implementation choice.

This behavior is specific to `_Enum` because it is otherwise not an implementation choice whether or not a certain class exists.

> [we could] _formally_ specify that the _EnumImpl_ class should be ignored completely for **Up** computation

We could do that. However, it might actually be simpler to change the analyzer to take `_Enum` into account, following the actual implementation. This would eliminate the discrepancy between the analyzer and the CFE, and it is consistent with the feature specification. (Also, I think it's a good idea to avoid using different approaches in the specification and in the implementation(s) whenever possible.)

In addition to this we also have the other discussion, about the observability of private classes in a hierarchy:

> Or, short: _depth_ is part of your public API, and it shouldn't be.

I think it would be a rather substantial undertaking to make it unobservable that public classes (or classes from the current library which may be private or public) have superinterfaces that are private. We can, essentially, inspect the superinterface hierarchy by observing the response from tools like the analyzer about subtype relationships.

Next, it gets complicated if those private superinterfaces add new members (or override existing ones with a new signature) &mdash; in this case the developer _needs to know_ that those new members (or new signatures) exist, and it will easily get out of hand if we start to pretend that those declarations are located in one/some of the public classes in the hierarchy. For instance, it would be weird if we pretend that a private class from the current library doesn't have the declarations that we can actually see in the source code.

As usual, I think we should admit that the world has a particular structure (including: this class has a superinterface which is a private class), rather than trying to stitch up a picture that might seem "more appropriate" based on some abstract principles.

**Comment by lrhn** [comment]
The actual type hierarchy is:
```mermaid
graph BT;
    E1["E1"] --> EM1["_Enum&M_1"]
    E2 --> EM2["_Enum&M_2"]
    E1 --> I
    E2 --> I
    EM1 --> Enum_["_Enum"]
    EM1 --> M
    EM2 --> Enum_
    EM2 --> M
    M --> Enum
    Enum_ --> Enum
    Enum --> Object
    Enum_ ~~~ I
    I --> Object
```
As you can see, the intersection (everything above the two lowest levels) has two types at each level before `Object`.
That's why the "legacy LUB" (last step of **Up**) algorithm refuses to choose one of them, and falls back to `Object` as the only "most specific and also unambiguous" choice.

And that is why I'd recommend removing the inaccessible `_Enum` from the set before trying to find an upper bound, which would have allowed the algorithm to find `M`.
We don't have to change the depth of anything, just don't count the `_Enum`.
(I'd like to not count the inaccessible member towards depth, and I'd also ignore anonymous mixin applications, but that's not needed for this case.)

**Comment by eernstg** [comment]
Thanks for spelling out an algorithm idea, @hamsbrar!

I should say that we have to be very careful about changing algorithms like this one, because it is used during type inference and promotion and typing in general, and any change is likely to be breaking in a way that is pervasive (because type inference etc. are used everywhere), implicit (especially type inference, but also promotion), and subtle (say, the expression `[]` is inferred as `<num>[]`, but after the update to the LUB algorithm, or promotion, or type inference, etc., it's inferred as `<int>[]`; then a billion instructions later we encounter a dynamic error because someone attempts to add a `double` value to this list, or maybe it causes a logical bug because a different branch is taken in `if (myList is List<int>) ...`). So the discussion about LUB or about any particular algorithm isn't going to end soon. ;-)

A few things to note: 

With `extends T1, T2...Ti` we know that `i` is one because it isn't possible to have more than one syntactic superclass (and we can only have zero when `extends` is omitted, too, in which case we have `extends Object` implicitly).

The repeated application of step 3 (that is, step 3 and 4) will then yield a set of superinterfaces of `A` which is a subset of the ones that are gathered by the current algorithm (which includes all superinterfaces, including all mixin applications).

Step 3 and 4 will terminate unless the superinterface graph is cyclic (which is an error in its own right), so there is no termination issue for these steps.

The reference to `E1` < `M` < `I` < `Enum` < `_Enum` < `Object` as a hierarchy seems to imply that `Enum` is a subtype of `_Enum` (which is not correct), but also that `I` is a subtype of `Enum` (which isn't correct), and `M` is a subtype of `I` (also not correct). It is also possible that it doesn't imply anything about the subtyping relationships: It is just a set of superinterfaces of `A` which is ordered by some ordering relation _R_.

One possible interpretation of that ordering relation _R_ is that it is determined by a breadth first traversal of the declared superinterfaces of `A` (with that, it makes sense that `M` occurs just before `I`). We would then simply ignore any superinterface when we encounter it for the 2nd, 3rd, etc. time during the traversal. I can't know if this is the way it was done, but it is at least consistent with the declarations in the original posting.

However, that ordering relation is radically different from the ordering which is used today (which is the 'depth' of each of the superinterfaces), and it can actually contradict the subtyping relationship:

```dart
class D {}
class E extends D {}
class F implements E, D {}
class G1 extends Object implements F {} // "Object < F < E < D".
class G2 extends Object implements F {} // "Object < F < E < D".
```

So `LUB(G1, G2)` would then be `Object`, and the hierarchy in the comments further seem to imply that `Object` is a subtype of all the other types, and `E` is a subtype of `D` (both of which are false). Actually, we'd probably want `LUB(G1, G2) == F` (which is also the result that we get today). We could also have a `class F1 implements E, D {}` and a `class F2 implements D, E {}`, and then the hierarchies would contradict each other.

I think this illustrates that there are many opportunities to get this wrong. In any case, do keep an eye on https://github.com/dart-lang/language/issues?q=is%3Aopen+is%3Aissue+label%3Aleast-upper-bound in order to see where this goes.

**Comment by lrhn** [comment]
The issue that the legacy LUB is trying to solve is, given two types, which can each have an arbitrary (and infinite) number of supertypes, choose one which is a supertype of both, with an optimization goal of making it as "close" to the original types as possible. And do so in a *deterministic*, *consistent*, and *efficiently computable* way.

The first step it takes is to ignore all the supertypes, and focus only on the shared super-interfaces. A `List<int>` implements `Iterable<int>`, but it also has `Iterable<num>` and `List<num>` as supertypes. All types have an infinity of supertypes (if nothing else `FutureOr<T>`, `FutureOr<FutureOr<T>>`, etc. is one such set, and `num`, `Comparable<num>`, `Comparable<Comparable<num>>` is another.), but their set of superinterfaces is finite.
(And that's why the LUB of `List<num>` and `Iterable<int>` is not `Iterable<num>`, because `Iterable<num>` is not a superinterface of `List<num>`. Maybe restricting to just superinterfaces it's too limiting, and we could get better results if we didn't, but it successfully reduces an infinite problem to a finite problem.)

Then, given the set of shared superinterfaces, it tries to find a measure for "closest" to the original types, that it can optimize for. That was chosen, semi-arbitrarily, as the maximal depth of the type in the super-interface graph, rooted at `Object` with depth zero. (`Null` is handled before we get here, so it never occurs in this algorithm, otherwise it too would have had to have depth zero.)

Then it orders the potential results by their depth, and starting from the deepest potential solutions, it checks whether that depth level has a *single* type. If it does, it's the chosen type. If not, the entire level is skipped.
**Why is that?** 

The reason to skip the level is to avoid making an *arbitrary* choice, because "arbitrary" goes against the goal of "predictable".
The algorithm could have said: "Choose a random type from that level". That would be bad *because* it would make the algorighm both unpredictable and inconsistent. Two compilations of the same program can yield different types. Two UP operations on the same types can give different types, so `var x = b ? e1 : e2; x = b ? e1 : e2;` could, sometimes, refuse to compile, because it chooses two different unrelated types as the upper bound of the same two types, in two different places. (OK, random is *obviously* bad, it's like the optimally bad choice for resolving ambiguity, so no surprise here.)

Your design here is deterministic, but it still suffers from choosing *arbitrarily*, based on properties that should not have semantic effect.
* The order of the initial list of types, e.g., `HierarchyA`, depends on the *syntactic* order of interfaces. If you change `implements I, J` to `implements J, I`, it may change the type of an expression somewhere else in a program you've never heard of. That makes the order of your implements clauses part of your public API, something that others might rely on, and where changing it is potentially breaking. That's bad. Not "random" bad, but pretty damning.
* The two steps that has "Terminatie with this item as result" are ordered, with the first type being checked first. That means that changing `b ? e1 : e2` to `!b ? e2 : e1` can change the type of the result. That's also a completely valid rewrite, that shouldn't have any effect on anything.

So, if we were to just pick a type that both share, and do it *predictably*, *consistently* and not affected by arbitrary syntactic choices, instead of using random to choose, the original algorithm could say:
* Look at the deepest non-empty level of common subtypes. Choose the type whose declaration's name is first alphabetically (prefer public over privat). If two types have the same name, tie-break by the alphabetical ordering of their declaring library's URI.

That's completely deterministic, it doesn't depend on how things are declared syntactically, it doesn't depend on the order of the operands to the LUB operation. It's main issue is that it's arbitrary. And, heck, maybe that's OK. Sometimes it's better to get a result which is correct, and predictable, but not obviously the only possible choice, than to get a result which is `Object`, just because there were two better alternatives and we were afraid to choose.

We can filter out inaccessible types first, if we want to avoid havingt them as result. (I'd want that.)

(And it still cannot choose an anonymous mixin application, because any type which has an anonymous mixin application as superinterface will also have its singular subtype as superinterface, so we'll always find that instead.)

The biggest issue here is that it relies on *depth*, which itself depends on implementation choices which shouldn't be publicly visible. If you change `class Foo {...}` to `abstract class _FooBase {...} class Foo extends _FooBase {...}`, which should be a completely invisible change to other libraries, you change the depth of `Foo`, which may change the type of an expression in some program that you've never heard of. A simple refactoring becomes a breaking change for someone else. That's bad, and if we are going to change LUB, and potentially break some existing code, I'll want to change it to something that doesn't have *this* abstraction leaking property either.

@eernstg 
> As usual, I think we should admit that the world has a particular structure (including: this class has a superinterface which is a private class), rather than trying to stitch up a picture that might seem "more appropriate" based on some abstract principles.

I strongly disagree, and believe that ensuring that *abstraction* actually *works*, and prevents depending on implementation details behind the abstraction, is a cornerstone of providing a solid foundation for programmers.

The "public API" that I keep talking about is the totality of what your library is promising its clients. Anything that they can *depend on* in such a way that changing it is breaking. The types, their subtyping behavior, their member signatures, and their dynamic behavior are all part of that.

The ability to add `base`, `final` and `interface` is *taking away* things from the public API. 
If an implementation detail leaks into the public API, it's no longer an implementation detail, and then it becomes something the author has to deliberately design for and around.

Every part of your public API should be *deliberate*. 
The language doesn't just allow something because it can, we never automatically make something `const` just because it could be, because it's a promise that it would be breaking to change away from.

And for that to be something authors can actually work with, it needs to be *predictable*.
It should be clear which parts of a library is part of its public API, and what is not. Which things you can safely change, and which you cannot.

We generally say that what's inside a method body is invisible from the outside. That's why `async`, `async*` and `sync*` are to the right of the declaration, not on the left. It's an implementation choice for that particular member body, not part of the public signature. 
And why we have getters and setters, rather the fields, as part of the public API, so that you can safely change a field to a getter/setter and back without affecting clients. And therefore no client behavior must ever depend on whether a getter comes from a field or an explicit declaration. It's equally impossible to see whether a field is `late` or not, because it's all getters and settes.

A public class exposes its interface (set of member signatures), its static declarations, its constructors (and even whether some of them are generative, if the class can be extended) and its *public* superinterfaces.

There is, or *should be*, no way for other libraries to detect the *private* superinterfaces.
If you cannot name a type, it might as well not exist. At runtime, all you can is object-instance-of-type and type-subtype-of-type checks, and `==` on `Type` objects. 
If you cannot name a type, it might as well not exist. For all practical purposes, your code should work just as if that type didn't exist, or at least as if it doesn't have any property


A library can break its own abstraction.
* Have a public type alias for a private type.
* Have a public function signature mentioning a private type.
* Have a public type which mentions a private type (like `implements Iterable<_Private>`).
* Expose instances of a private class which has members or superinterfaces that are not part of the public static type.

But if it *doesn't*, it should be able to rely on private types being invisible to other libraries, even their position in the type hiererachy.
We should not be exporting the *type hierarchy*, just the *superinterface* relation. The complete type hierarchy is a red herring, that is, itself, an implementation detail of our type system. It contains details that are implementation details to individual libraries.

And *depth* breaks that abstraction. And it shouldn't. (Or we should stop using it.)

**Comment by lrhn** [comment]
> such anomalies as **UP** having different results in different libraries when invoked on the same arguments:

I want **UP**  to depend on the context type too, so depending on the library is small stuff in comparison. 😁 

As long as the superinterfaces do not contain any of your own library's private declarations, the behavior of such an **UP** would perfectly predictable. And if it does contain some of your own library's private declaration, you'd be the one to know, and care, and not be surprised by their existence.
I can't come up with a realistic scenario where **UP** giving you a type which is private to another library is what you *want*. I, and users, keep coming up with cases where it *isn't* what you want, but you get it anyway.
Consistently giving a useless result isn't *better*.

We can't know what users intend. But we can make generalized assumptions. A type being declared private is 99.9% of the time a type that other libraries shouldn't know about. Otherwise, why go to the effort of making it privately named at all.
Sure, there might be exceptions, but that's what they are.

And absent of a better hint, I think we should use that as a hint that giving such a type as the result of **UP** in another library, is giving a useless result. And we should give a better result instead.

**Comment by eernstg** [comment]
@lrhn and I had a long discussion about this IRL. We still don't agree, but the discussion clarified some overall relevant points.

The main reason why I want to use an `UP` (last case, that is, the Dart 1 algorithm) that does _not have an exception_ for private type introducing declarations (`class`, `mixin`, `mixin class`, `extension type`, plus whatever comes next) is that it is inconsistent.

If we do that then we should also do the following in order to be consistent:

- It should be an error for a public API to refer to a private type (for example, no return type like `_Secret?`, no parameter type like `FutureOr<_Secret>`, no superinterfaces like `extends _Secret with _SecretMixin implements _SecretInterface`). It is probably benign for a private method to have a private return type etc, but we'd need to check that kind of situation carefully.
- Inference of an expression should never yield type arguments or declared types that are or contain private types from a different library. (So we get the "nearest" non-private supertype or `Object?` or what?)
- It should be an error for a private type introducing declaration to declare a public member, or to override a public member with a declaration whose signature differs from the combined member signature of its superinterfaces.
- It might be an error for a public type alias to refer to a private type (like `typedef Foo = _Foo<int>;`).
- We could make it a run time error to invoke `e.runtimeType` in the case where the run-time type of the value of `e` is a private type from a different library.
- There are probably additional situations where a private type "leaks" into a different library.

We could introduce the ability for a mixin to have no type of its own. For every member declaration _D_ it would then be forced to have superinterfaces such that there is a combined member signature with the same name as _D_ and with the same types everywhere (same return type, no covariant specialization, same parameter types, etc). A mixin with no type could reasonably be ignored during the computation of `UP` (a bit like an "epsilon step" in some formalisms), which seems to be exactly the idea that "we just delete certain types before we choose the result to return from `UP`".

But we _don't do that_, and it would be a potentially massively breaking change. Moreover, the restriction on type aliases will gratuitously destroy a significant amount of expressive power. Surely there will be some useful designs relying on not having each of the other restrictions.

I'd very much prefer to have lints to support every developer who wishes to enforce these restrictions as far as possible. We already have [library_private_types_in_public_api](https://dart.dev/tools/linter-rules/library_private_types_in_public_api), and we could easily add an extra lint for public members of private classes, for inferred variable types which are private to a different library, etc.

It should also be noted that if we have these restrictions in the language then the Dart 1 algorithm in `UP` will never encounter a case where there is a need to remove a private class from a superinterface graph, because they won't occur in such graphs in the first place. Private declarations can depend on each other, and they may have a non-private LUB, but they won't ever be arguments to `UP` in a different library because they aren't available via any public APIs (and they are not available by transitive traversals of superinterfaces, etc.). Otherwise, private declarations would be treated just like other declarations by `UP` in the library that declares them. So if we are consistent then we don't _need_ the special exception where `UP` deletes all the private types from other libraries (but we might do it with typeless mixins).

**In short**: Let's use simple and consistent rules, and treat private types in a way that resembles the treatment they get elsewhere (e.g., when `var x = ...;` gives `x` the type `_Secret`). For anyone who wishes to avoid private types from different libraries, let's give them good support for doing that by means of lints.

**Comment by lrhn** [comment]
And my general argument against is that **UP** is not about being consistent. It's a heuristic function that tries to give a *useful* upper bound, while being deterministic, symmetric, non-arbitrary and possibly efficient.

The "non-arbitrary" means that if there are two *equally valid* choices, it doesn't choose one of them. Not even if it could do it deterministically and symmetrically. And that might still be more of an artifact of being deterministic, symmetric, and fairky easily specified, than being an actual design choice.

But it's a heuristic, which worked adequately when it was designed. 
There is nothing fundamental about the definition of the *depth* function used, or using a depth at all. It's not trying to be *consistent* with anything, it's just trying, best effort, to find an upper bound that is useful to the user, and within the design constraints. 
It's *never* been a _good_ design for a least upper bound, but it comes from a time where there were implicit downcasts everywhere, and types were ignored entirely in production mode. It was *adequate* at the time.
It was not adequate for Dart 2. It wasn't massively worse for Dart 2.12 (null safety), but with switch expressions, the shortcomings of **UP** hits people on a daily basis.

And so, heuristically, I think we'd be giving the user a more *useful* result in the very vast majority of cases, if we omit types with inaccessible declarations from the set of possible results. (And any anonymouys mixin applications.)

Any type you can't name, is *more likely* to not be a type that you can use.

The examples here, of private types that are usefully inferred outside of their declared libraries, is code that would never pass code-review. It's *possible* to write such code. It's just not a good idea, except perhaps in a very minuscule amount of cases, and we should prevent ourselves from improving the majority of cases because of that.

And the *added benefit* would be that authors don't need to worry about their private declarations getting leaked in *that* way. They can still leak them in other ways, but that's on themselves. This was one way of leaking a private type that was mostly outside of the author's control, and not something lints would have much of a chance of detecting. Leaking them through your own API is much easier to detect.

(But I also think that a complete redesign, where we use the context type as well as the operand types, and go away from *depth* entirely, and possibly generalize to more than two operands, because **UP** is not associative, would probably be even better. In the design space, with *smaller* modifications to the existing **UP**, I'd remove the inaccssbile types that usually just get in the way of finding another result. If we allow bigger changes, the sky is the limit!)

**Comment by lrhn** [comment]
It is specified. Mixin applications are ordered, so that later mixin applications can do `super.foo` invocations on methods from earlier mixins. They create a chain of superclasses, which are each an anonymous mixin's application of one mixin on the previous class in the chain.
That's quite deliberate, no deity invocations needed.

**Comment by eernstg** [comment]
@hamsbrar wrote:
> if we want to support both users i.e a user who is NOT OK with an inaccessible type and a user who is, then I don't think adding a simple lint will help.

The lint isn't going to help a developer who is experiencing a typing which is an outright error or perhaps just an inconvenience, it is intended to help the _authors and maintainers_ of the types that we're computing an upper bound for.

The assumption is that those authors and maintainers will want to avoid having certain structures in their type hierarchies. For instance, they might well want to make sure that no public type is a subtype of any private type. If that's true then we don't have to worry about how to deal with public types whose `UP` turns out to be a private type because it basically never happens.

If that's daily life, and we have a ruleset with no exceptions, then it's not going to be hard to reason about cases where it _does_ happen after all (that is, some expressions have a static type which is private to another library). Such cases might turn out to be a useful design pattern, and we generally don't make anything an error in Dart just because we haven't thought of a useful way to use it, yet.

In client code, it is always possible to force a particular solution if `UP` doesn't return the result we need. For example, we could have a superinterface structure where it could never work (because no reasonable definition of `UP` would be able to disambiguate the given types), so we'd have to force a specific choice explicitly:

```dart
class I {}
class J {}
class C1 implements I, J {}
class C2 implements I, J {}

var b = true;

void main() {
  var x = b ? C1() : C2(); // Must yield `Object`, which we don't want!
  var y = b ? C1() : C2() as I; // Disambiguate explicitly, `y` gets type `I`.
}
```

> ... turn it into an internal discussion

Sorry about the delayed responses. We do have lots of internal discussions (as well), and things can take time to settle. But do continue to contribute to any and all Dart discussions here if and when you want to do so!

> I still don't get why I'd want to do a `super.f` in `M2` which is not related to `M1` in any way.

Superinvocations in mixins are useful when you want to create a behavior from parts:

```dart
class A {
  final String name;
  A(this.name);
  String greeting() => name;
}

mixin Hello on A {
  @override
  String greeting() => 'Hello, ${super.greeting()}';
}

mixin Nice on A {
  @override
  String greeting() => '${super.greeting()}, nice to see you!';
}

mixin Join on A {
  @override
  String greeting() => '${super.greeting()}! Come and have cup of coffee!';
}

class B1 = A with Hello, Nice;

class B2 = A with Hello, Join;

void main() {
  print(B1('John').greeting());
  print(B2('John').greeting());
}
```

The mixins don't have to depend directly on each other, they just need to be designed for contributing to a behavior that has multiple parts.

**Comment by eernstg** [comment]
Check out https://github.com/dart-lang/language/issues/3282. ;-)

**Comment by eernstg** [comment]
@hamsbrar, thanks for the kind words!

> `A` is a valid choice, just as `B2` is. Choosing `B2` is a purely arbitrary decision.

Point well taken! `A` and `B2` are unrelated types, and there is little justification for the assumption that `B2` is a more useful result. It is not arbitrary, technically, because it relies on the 'depth' of those types, but the depth of a type isn't directly connected to the relevance or usefulness of a typed for the given context.

On the other hand, the property of being relevant to the context doesn't appear to be algorithmic (unless it is expressed directly as a context type schema), so we probably can't even hope to eliminate the reliance on some property which is "arbitrary" in that sense.

I think this would be yet another reason to keep this issue in mind: https://github.com/dart-lang/language/issues/1618.

**Comment by eernstg** [comment]
At this time, `--enable-experiment=inference-update-3` has been enabled by default, which means that the typing of the branching expression (in this case it's `?:`, but it applies to switch expressions and other forms as well) will use the context type, and the reported inconvenience is gone.

In particular, this example is accepted with no errors with `dart` and `dart analyze` from c42fd69433ca0741fda5ecf82803aaa2fda24165:

```dart
enum E { a, b }
mixin M on Enum {}
abstract interface class I {}
enum E1 with M implements I { a1 }
enum E2 with M implements I { a2 }

List<Enum> f(E e) => switch (e) {
      E.a => E1.values,
      E.b => E2.values,
    };

void main() => f(E.a).forEach(print);
```

I'll close this issue because the fix is in the pipeline and will be available soon.

Note that the example used to be related to https://github.com/dart-lang/language/issues/3665 as well, but this is no longer true (because the context type overrules the upper bound computation). Anyway, that issue is being fixed as well.
